docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+
+Action items:
+* Split out of the current client package. (@krousey)
+* Convert to an interface (currently, it's a struct).
+
+### Mux layer
+
+The client muxer layer has a map of group/version to cached RESTClient, and
+knows how to construct a new RESTClient in case of a cache miss.
+
+Tentative name: MetaClient
+
+Action items:
+* Move client cache out of kubectl libraries into a more general home.
+
+### High-level: Individual typed
+
+Our current high-level client allows you to write things like
+`c.Pods("namespace").Create(p)`; we will insert a level for the group.
+
+That is, the system will be:
+`client.GroupName().NamespaceSpecifier().RESTAction()`
+
+The types returned/consumed by such functions will be e.g. api/v1, NOT the
+current version inspecific types.
+
+We will provide this structure for each version of each group. It is infeasible
+to do this manually, so we will generate this. The generator will accept both
+swagger and the ordinary go types.
+
+Typed clients will be constructabale given a MetaClient.
+
+Action items:
+* Move current typed clients into new directory structure (described below)
+* Finish client generation logic. (@lavalamp)
+
+#### High-level, typed: Discovery
+
+A `DiscoveryClient` is necessary to discover the api groups, versions, and
+resources a server supports. It's constructable given a RESTClient. It is
+consumed by both the MetaClient and users who want to iterate over groups,
+versions, or resources. (Example: namespace controller.)
+
+Action items:
+* Refactor existing functions to present a sane interface, as close to that
+  offered by the other typed clients as possible. (@caeserxuchao)
+
+### High-level: Client Sets
+
+Because there will be multiple groups with multiple versions, we will provide an
+aggregation layer that combines multiple typed clients in a single object. We do
+this to constrain the testing matrix.
+
+A client set is constructable given a MetaClient.
+
+An example:
+
+```go
+import (
+  api_v1 "k8s.io/kubernetes/pkg/client/typed/generated/v1"
+  ext_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generated/extensions/v1beta1"
+  net_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generated/net/v1beta1"
+  net_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generic"
+)
+
+type Client interface {
+  API() api_v1.Client
+  Extensions() ext_v1beta1.Client
+  Net() net_v1beta1.Client
+
+  GroupVersion(group, version string) generic.Client
+}
+```
+
+Note that a particular version is chosen for each group. It is a general rule
+for our API structure that no client need care about more than one version of
+each group at a time.
+
+This is the primary deliverable that people would consume. It is also generated.
+
+Action items:
+* This needs to be built. It will replace the ClientInterface that everyone
+  passes around right now.
+
+### High-level: Generic
+
+The generic client lets users consume apis which are not compiled into their
```

Comment:
Might be better to call this the dynamic Kube client, or the Kube swagger client.  Generic doesn't feel right for some reason